### PR TITLE
feat(private_dns_resolver): add private dns resolver module

### PR DESCRIPTION
**Pull request by filipefigueredo**
## Description

The main goal of this PR is to create the Azure private DNS resolver module.
## PR Checklist

Please ensure the following before submitting this PR:

- [X] You complied with the code style of this project.
- [X] You formatted all Terraform configuration files to a canonical format (Hint: `terraform fmt`).
- [X] You validated all Terraform configuration files (Hint: `terraform validate`).
- [X] You executed a speculative plan, showing what actions Terraform would take to apply the current configuration (Hint: `terraform plan`).
- [X] The documentation was updated (`README.md`, `CHANGELOG.md`, etc.).
- [X] Whenever possible, the dependencies were updated to the latest compatible versions.

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix.
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no interface changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes.
- [ ] Other... Please describe:

#### Creating first the VNET and subnets

1. Create a new tf file (main.tf for example) at the root of this repo in your working directory and add the code below:
```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.6"
    }
  }

  required_version = ">= 1.3.0, < 2.0.0"
}

provider "azurerm" {
  features {
  }
}

## Resource Group
module "resource_group" {
  source      = "./azure/resource_group"
  name        = "pr_project"
  location    = "westeurope"
  environment = "dev"

## Virtual Network example
module "common_network" {
  source              = "./azure/virtual_network"
  resource_group_name = module.resource_group.name
  name                = "vnet-dev-westeu-1000"
  address_spaces      = ["10.99.0.0/16"]
  environment         = "dev"
  subnets = [
    {
      name                                          = "orange"
      address_prefixes                              = ["10.99.10.0/16"]
      private_link_service_network_policies_enabled = false
      private_endpoint_network_policies_enabled     = false
      service_endpoints                             = ["Microsoft.EventHub", "Microsoft.Web", "Microsoft.Sql"]
      delegations                                   = ["Microsoft.ContainerInstance/containerGroups", "Microsoft.Web/serverFarms", "Microsoft.Databricks/workspaces"]
    },
    {
      name                                          = "apples"
      address_prefixes                              = ["10.99.11.0/16"]
      private_link_service_network_policies_enabled = false
      private_endpoint_network_policies_enabled     = false
      service_endpoints                             = ["Microsoft.EventHub"]
      delegations                                   = ["Microsoft.ContainerInstance/containerGroups"]
    },
    {
      name                                          = "lemons"
      address_prefixes                              = ["10.99.12.0/16"]
      private_link_service_network_policies_enabled = false
      private_endpoint_network_policies_enabled     = false
      service_endpoints                             = []
      delegations                                   = []
    },
    {
      name                                          = "snet-outboundprivatedns-001"
      address_prefixes                              = ["10.99.13.0/24"]
      private_link_service_network_policies_enabled = false
      private_endpoint_network_policies_enabled     = false
      service_endpoints                             = []
      delegations                                   = ["Microsoft.Network/dnsResolvers"]
    },
    {
      name                                          = "snet-inboundprivatedns-001"
      address_prefixes                              = ["10.99.14.0/24"]
      private_link_service_network_policies_enabled = false
      private_endpoint_network_policies_enabled     = false
      service_endpoints                             = []
      delegations                                   = ["Microsoft.Network/dnsResolvers"]
    },
  ]
}

```
2. Run the command `terraform init` to initialize the Terraform project.
3. Run the command `terraform plan` to preview the changes that will be made to your infrastructure.
4. Apply the plan to make the vnet and subnets available.
5. Add the following code:

```hcl
# private DNS resolver
module "private_dns_resolver" {
  source = "./azure/private_dns_resolver"

  name                     = "myresolver-001"
  resource_group_name      = module.resource_group.name
  location                 = module.location.name
  environment              = module.environment.name
  vnet_name                = module.virtual_network.name
  vnet_resource_group_name = module.resource_group.name
  inbound_endpoints = ["snet-inboundprivatedns-001"]
  outbound_endpoints = ["snet-outboundprivatedns-001"]
}

output "name" {
  value = module.private_dns_resolver.name
}

output "id" {
  value = module.private_dns_resolver.id
}

output "vnet_name" {
  value = module.private_dns_resolver.virtual_network_name
}

output "vnet_id" {
  value = module.private_dns_resolver.virtual_network_id
}

output "inbound" {
  value = module.private_dns_resolver.inbound_endpoints
}

output "outbound" {
  value = module.private_dns_resolver.outbound_endpoints
}
```

### Does this PR introduce a breaking change?

- [ ] Yes.
- [X] No.

### CODE REVIEW

**Review by pmmluis — DISMISSED** [review]
I left some comments, but overall it's good to be approved.
Nice work

File: azure/private_dns_resolver/main.tf [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  byte_length = 4
+  keepers = {
+    vnet_name = lower(each.value.subnet_name)
+  }
+}
+
+resource "azurerm_private_dns_resolver_inbound_endpoint" "inbound_endpoint" {
+  for_each = { for endpoint in var.inbound_endpoints : endpoint.subnet_name => endpoint }
+
+  name                    = "inedp-${each.value.subnet_name}-${random_id.inbound_endpoint[each.key].hex}"
```
[the remark this answers is not in the document]
Comment (filipefigueredo) [reply azure/private_dns_resolver/main.tf:37]: Good point. It's not necessary and I for the sake of the simplicity I'll add the name of the subnet instead

File: azure/private_dns_resolver/main.tf [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  virtual_network_id  = data.azurerm_virtual_network.vnet.id
+
+  lifecycle {
+    ignore_changes = [tags]
+  }
+}
+
+data "azurerm_subnet" "inbound_endpoint" {
+  for_each = toset(var.inbound_endpoints)
+
```
Comment (pmmluis) [inline azure/private_dns_resolver/main.tf:19]:
```suggestion
```

File: azure/private_dns_resolver/main.tf [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  name                 = each.value
+  virtual_network_name = var.vnet_name
+  resource_group_name  = var.vnet_resource_group_name
+}
+
+resource "azurerm_private_dns_resolver_inbound_endpoint" "inbound_endpoint" {
+  for_each = toset(var.inbound_endpoints)
+
```
Comment (pmmluis) [inline azure/private_dns_resolver/main.tf:27]:
```suggestion
```

File: azure/private_dns_resolver/main.tf [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  lifecycle {
+    ignore_changes = [tags]
+  }
+}
+
+data "azurerm_subnet" "outbound_endpoint" {
+  for_each = toset(var.outbound_endpoints)
+
```
Comment (pmmluis) [inline azure/private_dns_resolver/main.tf:43]:
```suggestion
```

File: azure/private_dns_resolver/main.tf [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  name                 = each.value
+  virtual_network_name = var.vnet_name
+  resource_group_name  = var.vnet_resource_group_name
+}
+
+resource "azurerm_private_dns_resolver_outbound_endpoint" "outbound_endpoint" {
+  for_each = toset(var.outbound_endpoints)
+
```
Comment (pmmluis) [inline azure/private_dns_resolver/main.tf:51]:
```suggestion
```

**Review by pmmluis — APPROVED** [review]

**Review by pmmluis — APPROVED** [review]

**Pull request merged**